src/Transform.jl

```diff
@@ -538,121 +530,117 @@ active_sources: The sources that are being optimized.    Only these sources'
-# custom bounds.    Or maybe it should be part of ElboArgs with one transform
-# per celestial object rather than a single object containing an array of
-# transforms.
-function DataTransform(bounds::Vector{ParamBounds};
-                       active_sources=collect(1:length(bounds)),
-                       S=length(bounds))
-    @assert length(bounds) == length(active_sources)
-    @assert maximum(active_sources) <= S
-    active_S = length(active_sources)
-
-    function from_vp!{NumType <: Number}(
-            vp::VariationalParams{NumType}, vp_free::VariationalParams{NumType})
-        S = length(vp)
-        @assert length(vp_free) == active_S
-        for si=1:active_S
-            s = active_sources[si]
-            vp_to_free!(vp[s], vp_free[si], bounds[si])
-        end
+function from_vp!{T}(dt::DataTransform, vp::VariationalParams{T}, vp_free::VariationalParams{T})
+    active_sources = dt.active_sources
+    bounds = dt.bounds
+    for i in eachindex(active_sources)
+        vp_to_free!(vp[active_sources[i]], vp_free[i], bounds[i])
     end
+    return nothing
+end
 
-    function from_vp{NumType <: Number}(vp::VariationalParams{NumType})
-            vp_free = Array{NumType, 1}[
-                zeros(NumType, length(ids_free)) for si = 1:active_S]
-            from_vp!(vp, vp_free)
-            vp_free
+function to_vp!{T<:Number}(dt::DataTransform, vp_free::FreeVariationalParams{T}, vp::VariationalParams{T})
+    active_sources = dt.active_sources
+    bounds = dt.bounds
+    for i in eachindex(active_sources)
+        free_to_vp!(vp_free[i], vp[active_sources[i]], bounds[i])
     end
+    return nothing
+end
 
-    function to_vp!{NumType <: Number}(
-            vp_free::FreeVariationalParams{NumType}, vp::VariationalParams{NumType})
-        @assert length(vp_free) == active_S
-        for si=1:active_S
-            s = active_sources[si]
-            free_to_vp!(vp_free[si], vp[s], bounds[si])
-        end
-    end
+function from_vp{T}(dt::DataTransform, vp::VariationalParams{T})
+    vp_free = fill(zeros(T, length(ids_free)), length(dt.active_sources))
+    from_vp!(dt, vp, vp_free)
+    vp_free
+end
 
-    function to_vp{NumType <: Number}(vp_free::FreeVariationalParams{NumType})
-            @assert(active_S == S,
-                    string("to_vp is not supported when active_sources is a ",
-                           "strict subset of all sources."))
-            vp = Array{NumType, 1}[
-                zeros(NumType, length(CanonicalParams)) for s = 1:S]
-            to_vp!(vp_free, vp)
-            vp
-    end
+function to_vp{T}(dt::DataTransform, vp_free::FreeVariationalParams{T})
+    n_active_sources = length(dt.active_sources)
+    @assert(n_active_sources == dt.S,
+            string("to_vp is not supported when active_sources is a ",
+                   "strict subset of all sources."))
+    vp = fill(zeros(T, length(ids_free)), n_active_sources)
```

Comment:
This could be the source of the problem too. In addition to creating reference to arrays,  `length(ids_free) != length(CanonicalParams)`. We had `length(CanonicalParams)` here before.